### ISSUE TITLE
Bracket highlighting is counterintuitive

### ISSUE THREAD

**Issue by ned2** [jonathanslenders/ptpython#120]
This might be specific to my configuration, but when I have parenthesis highlighting on and move my cursor to a parenthesis, the background colour of the cursor changes from white to dark blue while the matching parenthesis is highlighted a light purple. This is completely counter-intuitive for me as the changed contrast makes it feel like my cursor has jumped to the matching parenthesis. In this image the cursor is actually on the left paren but looks all the world like it's on the right paren.

![parens_highlighting](https://cloud.githubusercontent.com/assets/2831640/15597098/3e40f72c-2410-11e6-8621-8f49c9be4d9d.png)

I've had to disable parenthesis highlighting as using it just breaks my brain. I tried to look into where I could go about overriding the colours but did not get very far. Is it possible to customise this behaviour within `~/.ptpython/config.py`?

**Comment by jonathanslenders** [comment]
Hi @ned2,

Thanks for reporting this. What terminal are you using?

@takluyver @Carreau This is probably also True for IPython.
I'm going to do some more testing this evening. It should be possible to find a background color that works on all terminals.

Jonathan

**Comment by ned2** [comment]
The screenshot was taken in Guake Terminal. I also see the same issue on
Gnome Terminal.

**Comment by ned2** [comment]
What about just leaving the cursor background color unchanged -- so in my setup where the screenshot was taken, the cursor (which is on the left paren) would be backgrounded white (with the character color inverted)? This makes tracking the cursor simple. It's also how the emacs package smartparens does bracket highlighting.

**Comment by yyt16384** [comment]
Maybe matching brackets should use two different tokens instead of only one (`Token.MatchingBracket`).

**Comment by takluyver** [comment]
@jonathanslenders thanks for the heads up. We'll keep an eye on it, and we can always turn the default back off if it looks like it's going to be more trouble than it's worth. I'm enjoying it so far!

**Comment by ned2** [comment]
I just cloned the IPython master branch and I'm seeing the same confusing
highlighting. Perhaps this is not something that is a problem for everyone
or there could be terminal/environments where the colours are not a problem.

For me, it's especially hard to visually process when you have an opening
and a closing parenthesis with nothing between. When moving the cursor
across the two characters, I consistently think that my cursor is on the
other paren that it actually is and start writing text outside of the
parentheses.

**Comment by Carreau** [comment]
I think the confusion is because matching bracket have a bg color. 

I would be in favor of changing the default matching bracket style to something without backgroud:

```
    Token.MatchingBracket:                        '#00bbee bold',
```

For example seem to work great on light and dark, Bold seem necessary as brackets have often only few pixels, so bold increase the visual contrast at the edge of the vision. Underline can also be effective.

**Comment by ned2** [comment]
I agree this is part of the problem, but I think the matching bracket's bg
color changing is only a problem in conjunction with the current cursor bg
color changing to a dark color -- which makes the matching cursor look like
the current cursor. I think it would be sufficient to change the behavior
of the current cursor and leave the matching token's bg color the
untouched.

**Comment by jonathanslenders** [comment]
Hi all,

Yesterday, I did some experiments, and the following commit is the best that I could get. (With background color.) It looks good in gnome-terminal dark+light background, Guake, xterm black+white background:
https://github.com/jonathanslenders/python-prompt-toolkit/commit/3e17fdf2f2aa3da8778365f3a62b6bcae204c60b

In that commit, I also use a different token for a bracket that is under the cursor position. So, you can style it differently. I think I keep the style from that commit as the default in prompt-toolkit, but anyone should be able to override it. It is impossible to find anything that works everywhere. Even the defaults from Vim that I tried to copy don't work everywhere.

For ptpython: I have a second look later on.

Jonathan

**Comment by Carreau** [comment]
> I agree this is part of the problem, but I think the matching bracket's bg
> color changing is only a problem in conjunction with the current cursor bg
> color changing to a dark color -- which makes the matching cursor look like
> the current cursor. I think it would be sufficient to change the behavior
> of the current cursor and leave the matching token's bg color the
> untouched.

Ah, but that's a missunderstanding, ptk does not change the color of the cursor.
In most terminals the color of the cursor is just "inverted" with respect to the rest of the text:
![smc](https://cloud.githubusercontent.com/assets/335567/15635097/80551d16-258a-11e6-841d-db39ca0e633e.gif)

If you just remove the background color from the matched bracket,the cursor does not change:

![smc2](https://cloud.githubusercontent.com/assets/335567/15635104/a464b478-258a-11e6-9d27-1f5fdacdda5a.gif)

Agreed we could hack around and have a specific token type if cursor in **on** the bracket, but that seem overkill.

Note, the cursor being a box is also a choice of the terminal emulator, I can chose it to be underlined, or a vertical bar as well.

**Comment by ned2** [comment]
@Carreau ah right, yeah that makes much more sense -- the dark background color was just the inverted color of the matching parenthesis. 

In my current Emacs setup, I wind up with behaviour that looks the same as the Smart Cursor Color behaviour from iterm2. Not sure how Emacs achieves this, but seeing as this feels like it _just works_, it felt like ptpython/prompt_toolkit was doing something extra/weird.

@jonathanslenders ah fantastic, default colors are now much more comprehensible to me. Thanks!

**Comment by naggie** [comment]
I'd love to be able to disable this. 


![image](https://cloud.githubusercontent.com/assets/208440/20756200/aebee16c-b709-11e6-9979-ca6fddcb3ded.png)


Which one is the cursor?

**Comment by mixmastamyk** [comment]
What is the color-scheme key to customize this?  The default is too bringt on a dark background/light text.  I had to shut it off. 😞
